src/mono/mono/utils/lifo-semaphore.c

```diff
@@ -80,10 +88,238 @@ mono_lifo_semaphore_release (LifoSemaphore *semaphore, uint32_t count)
+			   LifoSemaphoreAsyncWaitWaitEntry *entry)
+{
+	if (*head == entry) {
+		*head = entry->next;
+	}
+	if (entry->previous) {
+		entry->previous->next = entry->next;
+	}
+	if (entry->next) {
+		entry->next->previous = entry->previous;
+	}
+}
+
+/* LOCKING: assumes semaphore is locked */
+static LifoSemaphoreAsyncWaitWaitEntry *
+lifo_js_find_waiter (LifoSemaphoreAsyncWaitWaitEntry *entry)
+{
+	while (entry) {
+		if (entry->state == LIFO_JS_WAITING)
+			return entry;
+		entry = entry->next;
+	}
+	return NULL;
+}
+
+static gboolean
+lifo_js_wait_entry_no_thread (LifoSemaphoreAsyncWaitWaitEntry *entry,
+			     pthread_t cur)
+{
+	while (entry) {
+		if (pthread_equal (entry->thread, cur))
+			return FALSE;
+		entry = entry->next;
+	}
+	return TRUE;
+}
+
+void
+mono_lifo_semaphore_asyncwait_prepare_wait (LifoSemaphoreAsyncWait *sem,
+				     int32_t timeout_ms,
+				     LifoSemaphoreAsyncWaitCallbackFn success_cb,
+				     LifoSemaphoreAsyncWaitCallbackFn timeout_cb,
+				     intptr_t user_data)
+{
+	mono_coop_mutex_lock (&sem->base.mutex);
+	if (sem->base.pending_signals > 0) {
+		sem->base.pending_signals--;
+		mono_coop_mutex_unlock (&sem->base.mutex);
+		success_cb (sem, user_data); // FIXME: queue microtask
+		return;
+	}
+
+	pthread_t cur = pthread_self ();
+
+	/* Don't allow the current thread to wait multiple times.
+	 * No particular reason for it, except that it makes reasoning a bit easier.
+	 * This can probably be relaxed if there's a need.
+	 */
+	g_assert (lifo_js_wait_entry_no_thread(sem->head, cur));
+
+	LifoSemaphoreAsyncWaitWaitEntry *wait_entry = g_new0 (LifoSemaphoreAsyncWaitWaitEntry, 1);
+	wait_entry->success_cb = success_cb;
+	wait_entry->timeout_cb = timeout_cb;
+	wait_entry->sem = sem;
+	wait_entry->user_data = user_data;
+	wait_entry->thread = pthread_self();
+	wait_entry->state = LIFO_JS_WAITING;
+        wait_entry->refcount = 1; // timeout owns the wait entry
+	wait_entry->js_timeout_id = emscripten_set_timeout (lifo_js_wait_entry_on_timeout, (double)timeout_ms, wait_entry);
+	lifo_js_wait_entry_push (&sem->head, wait_entry);
+	mono_coop_mutex_unlock (&sem->base.mutex);
+	return;
+}
+
+void
+mono_lifo_semaphore_asyncwait_release (LifoSemaphoreAsyncWait *sem,
+				uint32_t count)
+{
+	mono_coop_mutex_lock (&sem->base.mutex);
+
+	while (count > 0) {
+		LifoSemaphoreAsyncWaitWaitEntry *wait_entry = lifo_js_find_waiter (sem->head);
+		if (wait_entry != NULL) {
+			/* found one.  set its status and queue some work to run on the signaled thread */
+			pthread_t target = wait_entry->thread;
+			wait_entry->state = LIFO_JS_SIGNALED;
+			wait_entry->refcount++;
+			// we're under the mutex - if we got here the timeout hasn't fired yet
+			g_assert (wait_entry->refcount == 2); 
+			--count;
+			/* if we're on the same thread, don't run the callback while holding the lock */
+			emscripten_dispatch_to_thread_async (target, EM_FUNC_SIG_VI, lifo_js_wait_entry_on_success, NULL, wait_entry);
```

Comment:
is this sufficient to ensure the lock is released before it runs? i assume it posts it to the message queue of that thread?

Reply:
yea the `_async` variant is like `queueMicrotask`: it will not run the callback immediately, so we will get a chance to unlock.